File: app/src/main/java/net/mmho/photomap2/PhotoCursor.java
package net.mmho.photomap2;

import android.database.Cursor;
import android.database.CursorWrapper;
import android.provider.MediaStore;

import com.google.android.gms.maps.model.LatLng;

import net.mmho.photomap2.geohash.GeoHash;

import java.util.ArrayList;

class PhotoCursor extends CursorWrapper{

    final public static String[] projection = new String[]{
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.DISPLAY_NAME,
            MediaStore.Images.Media.LATITUDE,
            MediaStore.Images.Media.LONGITUDE,
            MediaStore.Images.Media.DATE_TAKEN,
            MediaStore.Images.Media.ORIENTATION,
    };

    private static final int HASH_CHARACTER_LENGTH=9;


    public PhotoCursor(Cursor cursor) {
        super(cursor);
    }

    long getID(){
        return getLong(getColumnIndexOrThrow(MediaStore.Images.Media._ID));
    }

    public LatLng getLocation(){
        float latitude = getFloat(getColumnIndexOrThrow(MediaStore.Images.Media.LATITUDE));
        float longitude = getFloat(getColumnIndexOrThrow(MediaStore.Images.Media.LONGITUDE));
        return new LatLng(latitude,longitude);
    }

    public GeoHash getGeoHash(int character){
        return GeoHash.createWithCharacterCount(getLocation(),character);
    }

    public HashedPhoto getHashedPhoto(int character){
        return new HashedPhoto(getID(),getGeoHash(character));
    }

    public ArrayList<HashedPhoto> getHashedPhotoList(){
        ArrayList<HashedPhoto> list = new ArrayList<>();
        if(isClosed()||!moveToFirst()) return list;
        do{
            list.add(getHashedPhoto(HASH_CHARACTER_LENGTH));
        }while(moveToNext());
        return list;
    }

}

File: app/src/main/kotlin/net/mmho/photomap2/PhotoCursor.kt
package net.mmho.photomap2

import android.database.Cursor
import android.database.CursorWrapper
import android.provider.MediaStore.Images.Media.*
import com.google.android.gms.maps.model.LatLng
import net.mmho.photomap2.geohash.GeoHash
import java.util.*

internal class PhotoCursor(cursor: Cursor) : CursorWrapper(cursor) {

    val id: Long
        get() = getLong(getColumnIndexOrThrow(_ID))

    val location: LatLng
        get() {
            val latitude = getFloat(getColumnIndexOrThrow(LATITUDE))
            val longitude = getFloat(getColumnIndexOrThrow(LONGITUDE))
            return LatLng(latitude.toDouble(), longitude.toDouble())
        }

    private val dateTaken:Long
        get() = getLong(getColumnIndexOrThrow(DATE_TAKEN))

    private fun getGeoHash(character: Int): GeoHash =
        GeoHash.createWithCharacterCount(location, character)

    private fun getHashedPhoto(character: Int): HashedPhoto =
        HashedPhoto(id, dateTaken, getGeoHash(character))

    val hashedPhotoList: ArrayList<HashedPhoto>
        get() {
            val list = ArrayList<HashedPhoto>()
            if (isClosed || !moveToFirst()) return list
            do {
                list.add(getHashedPhoto(HASH_CHARACTER_LENGTH))
            } while (moveToNext())
            return list
        }

    companion object {
        val projection = arrayOf(_ID, DATA, DISPLAY_NAME, LATITUDE, LONGITUDE, DATE_TAKEN, ORIENTATION)
        private val HASH_CHARACTER_LENGTH = 9
    }

}
